### ISSUE TITLE
[perf] benchmark and profiling ast-grep core matching logic

### ISSUE THREAD

**Issue by HerringtonDarkholme** [ast-grep/ast-grep#144]
https://medium.com/source-and-buggy/data-driven-performance-optimization-with-rust-and-miri-70cb6dde0d35

**Comment by 0xdevalias** [comment]
The tl;dr of that blog seemed to be that they first used `flamegraph`, which was mildly useful, but not specifically detailed enough:

- https://github.com/flamegraph-rs/flamegraph
  - > A Rust-powered flamegraph generator with additional support for Cargo projects! It can be used to profile anything, not just Rust projects!

And then moved on to using `miri` and a less-obvious feature of it to generate a far more detailed trace, that could then be analysed within Google Chrome's DevTools performance tab:

- https://github.com/rust-lang/miri
  - > An experimental interpreter for [Rust](https://www.rust-lang.org/)'s [mid-level intermediate representation](https://github.com/rust-lang/rfcs/blob/master/text/1211-mir.md) (MIR)
  - https://github.com/rust-lang/miri#miri--z-flags-and-environment-variables
    - > `-Zmiri-measureme=<name>` enables measureme profiling for the interpreted program. This can be used to find which parts of your program are executing slowly under Miri. The profile is written out to a file inside a directory called `<name>`, and can be processed using the tools in the repository https://github.com/rust-lang/measureme
      - https://github.com/rust-lang/measureme#crox
        - > `crox` turns `measureme` profiling data into files that can be visualized by the Chromium performance tools.
        - https://github.com/rust-lang/measureme/tree/master/crox#readme
